lib/src/json/builder/builder.dart

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import "../reader/reader.dart";
+
+/// A function reading JSON data from a [JsonReader] into a Dart value.
+///
+/// This is a general type which can be implemented
+/// for any type that can be built from JSON data.
+typedef JsonBuilder<T> = T Function(JsonReader);
+
+/// Parses a JSON value from [reader] into a plain JSON-like structure.
+///
+/// A JSON-like structure is either a number, a string, a boolean, `null`, or
+/// a `List<dynamic>` containing JSON-like structures, or a `Map<String, dynamic>`
+/// where the values are JSON-like structures.
+dynamic jsonValue(JsonReader reader) {
+  if (reader.checkObject()) {
+    return jsonObject(jsonValue)(reader);
+  }
+  if (reader.checkArray()) {
+    return jsonArray(jsonValue)(reader);
+  }
+  if (reader.tryNull()) return null;
+  return reader.tryNum() ??
+      reader.tryString() ??
+      reader.tryBool() ??
+      (throw FormatException("Reader has no value"));
+}
+
+/// Reads an integer from [reader].
+int jsonInt(JsonReader reader) => reader.expectInt();
+
+/// Reads a double from [reader].
+double jsonDouble(JsonReader reader) => reader.expectDouble();
+
+/// Reads a number from [reader].
+num jsonNum(JsonReader reader) => reader.expectNum();
+
+/// Reads a string from [reader].
+String jsonString(JsonReader reader) => reader.expectString();
+
+/// Reads a boolean from [reader].
+bool jsonBool(JsonReader reader) => reader.expectBool();
+
+/// Reads eiher a `null` or what [builder] reads from a [JsonReader].
+///
+/// If the next value of [builder] is `null`, then the result
+/// is `null`, otherwise returns the same value as [builder]
+/// on `reader`.
+///
+/// Can be used as, for example:
+/// ```dart
+/// var optionalJsonInt = jsonOptional(jsonInt);
+/// var optionalInt = optionalJsonInt(reader);  // int or null
+/// ```
+JsonBuilder<T /*?*/ > jsonOptional<T>(JsonBuilder<T> builder) =>
```

Comment:
**Just a comment**: I see a lot of nice methods in library and wonder if somehow these could be static methods on some class... It seems unreasonable to introduce an empty abstract class for namespacing alone.. But when coding these will be hard to discover, and in documentation they might drown.

I guess `jsonOptional` and `jsonArray` could be extension methods on `Builder<T>`, but this also seems a bit crazy :)

I don't have a good suggestion for improving discoverability. I'm merely noting that this is a lot of top-level methods that pollutes docs and is hard to discover (I guess the upside is that it might be quite readable).

Reply:
The problem with putting them inside a class is that you need to write the class name to access them. And it's against the style guide to have a static-only class (and putting functions into an otherwise unrelated class is just a hack around that rule).

The traditional approach would be to make these top-level functions and import the library with a prefix. However, this is an existing library/package which is not imported with a prefix. So I added the `json` prefix to each function instead.

No perfect solution available. 

(And yes, extensions on a function type is crazy, and doesn't work for the generic variants because the function types are not generic).